Add Workflow to Tag Issues Missing Area Labels

deepakrathore33 opened the pull request:
This PR introduces a scheduled and manually-triggerable GitHub Actions workflow to identify and tag open issues that are missing an Area-* label in the dotnet/roslyn repository.

## ✅ Key Features:
* Runs daily at 13:00 UTC and can be manually triggered.

* Filters open issues that:

  * Have the untriaged label

  * Do not have any Area-* label

  * Do not already have the `needs_area_label` tag

  * Do not have any assignee

* Adds the label `needs_area_label` to such issues

* Skips pull requests

* Skips workflows running on forks (restricted to dotnet/roslyn)

JoeRobich commented on .github/workflows/labeler-missing-area-label.yml:
```diff
@@ -0,0 +1,63 @@
+
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v3
```
Is it necessary to checkout the Roslyn source for this action?

deepakrathore33 replied:
Right, we don't need it

JoeRobich commented on .github/workflows/labeler-missing-area-label.yml:
```diff
@@ -0,0 +1,63 @@
+        uses: actions/checkout@v3
+
+      - name: Tag Issues Missing Area Labels
+        uses: actions/github-script@v6
```
Looks like the latest is [v7](https://github.com/actions/github-script) can we use it instead?

JoeRobich commented on .github/workflows/labeler-missing-area-label.yml:
```diff
@@ -0,0 +1,63 @@
+        with:
+          script: |
+            const AREA_LABEL_PREFIX = 'Area-';
+            const NEEDS_AREA_LABEL = 'needs_area_label';
```
Looks like we use `-` in our labels instead of `_`.

JoeRobich commented on .github/workflows/labeler-missing-area-label.yml:
```diff
@@ -0,0 +1,63 @@
+              if (issue.pull_request) continue;
+
+              const labels = issue.labels.map(label => label.name);
+              const hasUntriaged = labels.includes('untriaged');
```
Why do triaged issues not need an area label?

JoeRobich commented on .github/workflows/labeler-missing-area-label.yml:
```diff
@@ -0,0 +1,63 @@
+              const hasUntriaged = labels.includes('untriaged');
+              const hasAreaLabel = labels.some(name => name.startsWith(AREA_LABEL_PREFIX));
+              const hasNeedsAreaLabel = labels.includes(NEEDS_AREA_LABEL);
+              const hasAssignee = issue.assignee !== null;
```
Why do assigned issues not need an area label?

deepakrathore33 commented:
@JoeRobich I have updated the logic and removed the 'Interactive-*' labels from consideration. However, there are still about 100 [issues](https://github.com/dotnet/roslyn/issues?q=is%3Aissue%20state%3Aopen%20-label%3AArea-Compilers%20-label%3A%22Area-IDE%22%20-label%3A%22Area-Performance%22%20-label%3AArea-Interactive%20-label%3A%22Area-Language%20Design%22%20-label%3AArea-Infrastructure%20-label%3AArea-Analyzers%20-label%3AArea-IDE-Parent%20-label%3AInteractive-EnC%20-label%3AInteractive-Debugging%20%20-label%3AArea-External%20-label%3A%22Area-SDK%20and%20Samples%22&page=4) that will be affected.
